connector/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryDataWriterHelper.java

```diff
@@ -0,0 +1,364 @@
+
+    try {
+      this.writeStreamName = retryCreateWriteStream();
+    } catch (ExecutionException | InterruptedException e) {
+      throw new RuntimeException("Could not create write-stream after multiple retries.", e);
+    }
+    this.streamWriter = createStreamWriter(this.writeStreamName);
+    this.protoRows = ProtoBufProto.ProtoRows.newBuilder();
+  }
+
+  /**
+   * Submits a callable that creates a BigQuery Storage Write API write-stream to function
+   * {retryCallable}.
+   *
+   * @see this#retryCallable(Callable createWriteStream)
+   * @return The write-stream name, if it was successfully created.
+   * @throws ExecutionException If retryCallable failed to create the write-stream after multiple
+   *     retries.
+   * @throws InterruptedException If retryCallable was interrupted while creating the write-stream
+   *     during a retry.
+   */
+  private String retryCreateWriteStream() throws ExecutionException, InterruptedException {
+    return retryCallable(
+        () ->
+            this.writeClient
+                .createWriteStream(
+                    Storage.CreateWriteStreamRequest.newBuilder()
+                        .setParent(this.tablePath)
+                        .setWriteStream(
+                            Stream.WriteStream.newBuilder()
+                                .setType(Stream.WriteStream.Type.PENDING)
+                                .build())
+                        .build())
+                .getName());
+  }
+
+  /**
+   * A helper method in order to retry certain tasks: currently used for creating a write-stream,
+   * and finalizing a write-stream, if those requests reached a retriable error.
+   *
+   * @param callable The callable to retry.
+   * @param <V> The return value of the callable (currently, either a String (write-stream name) or
+   *     a FinalizeWriteStreamResponse.
+   * @return V.
+   * @throws ExecutionException If retryCallable failed to create the write-stream after multiple
+   *     retries.
+   * @throws InterruptedException If retryCallable was interrupted while creating the write-stream
+   *     during a retry.
+   */
+  private <V> V retryCallable(Callable<V> callable)
+      throws ExecutionException, InterruptedException {
+    DirectRetryingExecutor<V> directRetryingExecutor =
+        new DirectRetryingExecutor<>(
+            new RetryAlgorithm<>(
+                new ApiResultRetryAlgorithm<>(),
+                new ExponentialRetryAlgorithm(this.retrySettings, NanoClock.getDefaultClock())));
+    RetryingFuture<V> retryingFuture = directRetryingExecutor.createFuture(callable);
+    return directRetryingExecutor.submit(retryingFuture).get();
+  }
+
+  private StreamWriter createStreamWriter(String writeStreamName) {
+    try {
+      return StreamWriter.newBuilder(writeStreamName).build();
+    } catch (IOException | InterruptedException e) {
+      throw new RuntimeException("Could not build stream-writer.", e);
```

Comment:
Please change RuntimeException to a subclass with better description
nit: remove dot in the end of the message